embedded/servo_driver/servo_driver.hh

```diff
@@ -7,23 +7,24 @@
 
 #include <stdint.h> // uint8_t etc
 #include <string>
+#include <memory>
 
 #include "embedded/pwm_driver/pwm_driver.hh"
 #include "third_party/i2c/i2c.h"
 
 
 class ServoDriver {
 public:
-	ServoDriver(const int channel, PwmDriver &driver, const std::string &config_path);
+	ServoDriver(const int channel, const std::shared_ptr<PwmDriver> &driver, const std::string &config_path);
 	int init();
 	void set_percentage(int percentage);
 	int get_percentage() const;
 	void set_angle(float angle);
 
 private:
 	int channel_;
-        PwmDriver &pwm_driver_;
-        std::string config_path_;
+	const std::shared_ptr<PwmDriver> &pwm_driver_;
```

Comment:
This reference is still dangling, since it's only a reference to the shared pointer, it wasn't copied, so the reference counter didn't get incremented

you likely want:
`const std::shared_ptr<PwmDriver> pwm_driver_` (no ampersand)